pgx/src/array.rs

```diff
@@ -0,0 +1,235 @@
+    pub fn nulls(&self) -> Option<NonNull<[u8]>> {
+        // for expected behavior, see:
+        // postgres/src/include/utils/array.h
+        // #define ARR_NULLBITMAP
+        let len = (self.len() as usize + 7) >> 3; // Obtains 0 if len was 0.
+
+        // SAFETY: This obtains the nulls pointer, which is valid to obtain because
+        // the validity was asserted on construction. However, unlike the other cases,
+        // it isn't correct to trust it. Instead, this gets null-checked.
+        // This is because, while the initial pointer is NonNull,
+        // ARR_NULLBITMAP can return a nullptr!
+        let nulls = unsafe { slice_from_raw_parts_mut(pgx_ARR_NULLBITMAP(self.at.as_ptr()), len) };
+        // This code doesn't assert validity per se, but in practice,
+        // the caller will probably immediately turn this into a borrowed slice,
+        // opening up the methods that are available on borrowed slices.
+        Some(NonNull::new(nulls)?)
+    }
+
+    /// The slice of the data.
+    /// Oxidized form of ARR_DATA_PTR(ArrayType*)
+    ///
+    /// # Safety
+    ///
+    /// This is not inherently typesafe!
+    /// Thus you must know the implied type of the underlying ArrayType when calling this.
+    /// In addition, the raw slice is not guaranteed to be legible at any given index,
+    /// e.g. it may be an "SQL null" if so indicated in the null bitmap.
+    /// But even if the index is not marked as null, the value may be equal to nullptr,
+    /// thus leaving it correct to read the value but incorrect to then dereference.
+    ///
+    /// That is why this returns [`NonNull<[T]>`]: if it returned `&mut [T]`,
+    /// then for many possible types that can be **undefined behavior**,
+    /// as it would assert that each particular index was a valid `T`.
+    /// A Rust borrow, including of a slice, will always be
+    /// - non-null
+    /// - aligned
+    /// - **validly initialized**, except in the case of [MaybeUninit] types
+    /// It can be incorrect to assume that data that Postgres has marked "null"
+    /// follows Rust-level initialization requirements.
+    ///
+    /// [MaybeUninit]: core::mem::MaybeUninit
+    /// [`NonNull<[T]>`]: core::ptr::NonNull
+    pub unsafe fn data<T>(&mut self) -> NonNull<[T]> {
+        let len = self.len() as usize;
+
+        // SAFETY: Welcome to the infernal bowels of FFI.
+        // Because the initial ptr was NonNull, we can assume this is also NonNull.
+        // As validity of the initial ptr was asserted on construction of RawArray,
+        // this can assume the data ptr is also valid, allowing making the slice.
+        // This code doesn't assert validity per se, but in practice,
+        // the caller will probably immediately turn this into a borrowed slice,
+        // opening up the methods that are available on borrowed slices.
+        //
+        // Most importantly, the caller has asserted this is in fact a valid [T],
```

Comment:
You don't specifically document that it's asserting `[T]`, you're only asking for `T`, and there's some nuances between those.